chore: make configs more optional and mergeable

Gankra opened the pull request:
fixes #340

Gankra committed:
chore: make configs more optional and mergeable

fixes #340

manyinsects reviewed (approved):
we're defining `apply_layer` so many times that it really feels like it should be a trait, but that's not a blocker imo

Gankra commented:
I don't disagree but also I don't know what purpose the trait would actually serve -- I suppose streamlining the Option handling?

Gankra committed:
chore: introduce convenience methods for merging values

This introduces an ApplyLayer trait that already existed as a convention, as well as several
extension traits and blanket implementations to add various combinations of merging T/Option/BoolOr

ashleygwilliams commented on src/config/artifacts.rs:
```diff
@@ -36,6 +38,14 @@ pub struct Artifacts {
+impl ApplyLayer for Artifacts {
+    fn apply_layer(&mut self, layer: Self) {
+        self.cargo_dist.apply_opt(layer.cargo_dist);
+        // FIXME: should this get merged with e.g. `extend?`
```
where is the other location that a user can specify package_managers?

Gankra replied:
Nowhere, but each layer """has to""" assume that other layers can set these values, as a matter of like, robust abstraction / consistency?

Gankra replied:
These FIXMEs are more like "hey this code might make your sad if we add new layers that touch these values".

ashleygwilliams commented on src/config/mod.rs:
```diff
@@ -77,107 +77,39 @@ impl Config {
+            self.repository.apply_opt(custom.repository);
+            self.analytics.apply_layer(custom.analytics);
+            // FIXME: should this get merged with e.g. `extend?`
+            self.additional_pages.apply_opt(custom.additional_pages);
```
where is the other place that additional pages could be specified?

Gankra replied:
ditto

ashleygwilliams commented on src/config/oranda_config.rs:
```diff
@@ -30,23 +40,39 @@ pub struct MdBookConfig {
+}
+impl StyleConfig {
+    /// Get the theme
+    pub fn theme(&self) -> Theme {
```
why are we adding functions here instead of using the default values as values?

Gankra replied:
This is the crux of a discussion I brought up with @shadows-withal last week.  Some values in Theme should be Optional in OrandaConfig, but aren't Optional in the final Config. This raises a question: "Do we want the final Config to use the same struct for things like StyleConfig that OrandaConfig uses?" (Before my PR the answer was "yes" and after my PR it remains "yes")

* If no, then we can just have different Optionality and everything is clean, but now you are defining ~everything multiple times
* If yes, then users of the final Config need to deal with the fact that the value is Optional, even though it isn't. I chose to deal with this fact by removing the default value from the Default impl, and moving it to a method that wraps them up in a method that returns the default value.

ashleygwilliams reviewed (commented):
a few questions - 

in some places, you comment about extending vs clobbering- but it's not clear to me how a user is specifying those keys in different places... (they are only available in oranda.json as far as i know). my personal preference is clobber because it keeps the source of truth singular fo r each key and allows you to "turn off" things that may be specified in other files that you don't have full control over and/or want oranda to ignore. i am a little concerned though because in general those keys should already only be specified once.

also not clear why theme and syntax_theme are functions now- for a few reasons i think we should treat them as values. curious to understand more

ashleygwilliams commented:
i am gonna merge this to avoid conflicts, but @Gankra it'd be good to handle those FIXMEs on monday
